File: examples/go/player/agg/handlers/state.go
// Package handlers implements player aggregate command handlers.
//
// DOC: This file is referenced in docs/docs/examples/aggregates.mdx
//      Update documentation when making changes to StateRouter patterns.
package handlers

import (
	"encoding/hex"

	angzarr "github.com/benjaminabbitt/angzarr/client/go"
	pb "github.com/benjaminabbitt/angzarr/client/go/proto/angzarr"
	"github.com/benjaminabbitt/angzarr/client/go/proto/examples"
)

// PlayerState represents the current state of a player aggregate.
type PlayerState struct {
	PlayerID          string
	DisplayName       string
	Email             string
	PlayerType        examples.PlayerType
	AIModelID         string
	Bankroll          int64 // In smallest unit (chips)
	ReservedFunds     int64
	TableReservations map[string]int64 // table_root_hex -> amount
	Status            string
}

// NewPlayerState creates a new empty player state.
func NewPlayerState() PlayerState {
	return PlayerState{
		TableReservations: make(map[string]int64),
	}
}

// Exists returns true if the player has been registered.
func (s PlayerState) Exists() bool {
	return s.PlayerID != ""
}

// AvailableBalance returns the available balance (bankroll - reserved).
func (s PlayerState) AvailableBalance() int64 {
	return s.Bankroll - s.ReservedFunds
}

// IsAI returns true if this is an AI player.
func (s PlayerState) IsAI() bool {
	return s.PlayerType == examples.PlayerType_AI
}

// Event applier functions for StateRouter

// docs:start:state_router
func applyRegistered(state *PlayerState, event *examples.PlayerRegistered) {
	state.PlayerID = "player_" + event.Email
	state.DisplayName = event.DisplayName
	state.Email = event.Email
	state.PlayerType = event.PlayerType
	state.AIModelID = event.AiModelId
	state.Status = "active"
	state.Bankroll = 0
	state.ReservedFunds = 0
}

func applyDeposited(state *PlayerState, event *examples.FundsDeposited) {
	if event.NewBalance != nil {
		state.Bankroll = event.NewBalance.Amount
	}
}

func applyWithdrawn(state *PlayerState, event *examples.FundsWithdrawn) {
	if event.NewBalance != nil {
		state.Bankroll = event.NewBalance.Amount
	}
}

func applyReserved(state *PlayerState, event *examples.FundsReserved) {
	if event.NewReservedBalance != nil {
		state.ReservedFunds = event.NewReservedBalance.Amount
	}
	if event.TableRoot != nil && event.Amount != nil {
		tableKey := hex.EncodeToString(event.TableRoot)
		state.TableReservations[tableKey] = event.Amount.Amount
	}
}

func applyReleased(state *PlayerState, event *examples.FundsReleased) {
	if event.NewReservedBalance != nil {
		state.ReservedFunds = event.NewReservedBalance.Amount
	}
	if event.TableRoot != nil {
		tableKey := hex.EncodeToString(event.TableRoot)
		delete(state.TableReservations, tableKey)
	}
}

func applyTransferred(state *PlayerState, event *examples.FundsTransferred) {
	if event.NewBalance != nil {
		state.Bankroll = event.NewBalance.Amount
	}
}

// stateRouter is the fluent state reconstruction router.
var stateRouter = angzarr.NewStateRouter(NewPlayerState).
	On(applyRegistered).
	On(applyDeposited).
	On(applyWithdrawn).
	On(applyReserved).
	On(applyReleased).
	On(applyTransferred)

// docs:end:state_router

// RebuildState rebuilds player state from event history.
func RebuildState(eventBook *pb.EventBook) PlayerState {
	if eventBook == nil {
		return NewPlayerState()
	}

	// Start from snapshot if available
	if eventBook.Snapshot != nil && eventBook.Snapshot.State != nil {
		if eventBook.Snapshot.State.MessageIs(&examples.PlayerState{}) {
			var snapshot examples.PlayerState
			if err := eventBook.Snapshot.State.UnmarshalTo(&snapshot); err == nil {
				state := applySnapshot(&snapshot)
				// Apply events since snapshot
				for _, page := range eventBook.Pages {
					if page.Event != nil {
						stateRouter.ApplySingle(&state, page.Event)
					}
				}
				return state
			}
		}
	}

	return stateRouter.WithEventBook(eventBook)
}

func applySnapshot(snapshot *examples.PlayerState) PlayerState {
	bankroll := int64(0)
	if snapshot.Bankroll != nil {
		bankroll = snapshot.Bankroll.Amount
	}
	reservedFunds := int64(0)
	if snapshot.ReservedFunds != nil {
		reservedFunds = snapshot.ReservedFunds.Amount
	}

	reservations := make(map[string]int64)
	for k, v := range snapshot.TableReservations {
		reservations[k] = v
	}

	return PlayerState{
		PlayerID:          snapshot.PlayerId,
		DisplayName:       snapshot.DisplayName,
		Email:             snapshot.Email,
		PlayerType:        snapshot.PlayerType,
		AIModelID:         snapshot.AiModelId,
		Bankroll:          bankroll,
		ReservedFunds:     reservedFunds,
		TableReservations: reservations,
		Status:            snapshot.Status,
	}
}

File: examples/python/player/agg/handlers/state.py
"""Player state management for router pattern.

Provides PlayerState dataclass and StateRouter for event sourcing.
"""

from dataclasses import dataclass, field

from google.protobuf.any_pb2 import Any as AnyProto

from angzarr_client.proto.examples import player_pb2 as player_proto
from angzarr_client.state_builder import StateRouter


@dataclass
class PlayerState:
    """Player aggregate state."""

    player_id: str = ""
    display_name: str = ""
    email: str = ""
    player_type: int = 0
    ai_model_id: str = ""
    bankroll: int = 0
    reserved_funds: int = 0
    table_reservations: dict = field(default_factory=dict)
    status: str = ""

    @property
    def exists(self) -> bool:
        return bool(self.player_id)

    @property
    def available_balance(self) -> int:
        return self.bankroll - self.reserved_funds


# docs:start:state_router
def _apply_registered(state: PlayerState, event: player_proto.PlayerRegistered) -> None:
    state.player_id = f"player_{event.email}"
    state.display_name = event.display_name
    state.email = event.email
    state.player_type = event.player_type
    state.ai_model_id = event.ai_model_id
    state.status = "active"
    state.bankroll = 0
    state.reserved_funds = 0


# Why events carry final state (not deltas):
# Events contain new_balance (the result) rather than delta (amount deposited).
# This provides: (1) idempotent replay, (2) auditable history, (3) simpler appliers.
def _apply_deposited(state: PlayerState, event: player_proto.FundsDeposited) -> None:
    if event.new_balance:
        state.bankroll = event.new_balance.amount


def _apply_withdrawn(state: PlayerState, event: player_proto.FundsWithdrawn) -> None:
    if event.new_balance:
        state.bankroll = event.new_balance.amount


def _apply_reserved(state: PlayerState, event: player_proto.FundsReserved) -> None:
    if event.new_reserved_balance:
        state.reserved_funds = event.new_reserved_balance.amount
    table_key = event.table_root.hex()
    if event.amount:
        state.table_reservations[table_key] = event.amount.amount


def _apply_released(state: PlayerState, event: player_proto.FundsReleased) -> None:
    if event.new_reserved_balance:
        state.reserved_funds = event.new_reserved_balance.amount
    table_key = event.table_root.hex()
    state.table_reservations.pop(table_key, None)


def _apply_transferred(
    state: PlayerState, event: player_proto.FundsTransferred
) -> None:
    if event.new_balance:
        state.bankroll = event.new_balance.amount


player_state_router = (
    StateRouter(PlayerState)
    .on(player_proto.PlayerRegistered, _apply_registered)
    .on(player_proto.FundsDeposited, _apply_deposited)
    .on(player_proto.FundsWithdrawn, _apply_withdrawn)
    .on(player_proto.FundsReserved, _apply_reserved)
    .on(player_proto.FundsReleased, _apply_released)
    .on(player_proto.FundsTransferred, _apply_transferred)
)
# docs:end:state_router


def build_state(state: PlayerState, events: list[AnyProto]) -> PlayerState:
    """Apply events to state and return updated state."""
    for event_any in events:
        player_state_router.apply_single(state, event_any)
    return state


__all__ = ["PlayerState", "build_state", "player_state_router"]
